pkg/arvo/lib/lm.hoon

```diff
@@ -0,0 +1,702 @@
+::
+::::  Matrix type in single-precision floating-point @rs
+  ::
+/+  *lv
+|%
+++  lms
+  ^|
+  ~%  %lms  ..is  ~
+  |_  r/$?($n $u $d $z)   :: round nearest, round up, round down, round to zero
+  ::
+  ::  Manipulators
+  ::
+  ::    Cantor pairing function
+  ::      This is used so that the dimensions can be stored as a unique number.
+  ++  cantor
+    ~/  %cantor
+    |=  [m=@ud n=@ud]  ^-  @ud
+    (add (div (mul (add m n) (add m (add n 1))) 2) n)
+  ++  decantor
+    ~/  %decantor
+    |=  [z=@ud]  ^-  (list @ud)
+    =/  w  (div (sub p:(sqt (add (mul 8 z) 1)) 1) 2)
+    =/  t  (div (add (mul w w) w) 2)
+    =/  y  (sub z t)
+    =/  x  (sub w y)
+    ~[x y]
+  ::
+  ::    Zeroes
+  ++  zeros
+    ~/  %zeros
+    |=  [m=@ud n=@ud]  ^-  @lms
+    ~_  leaf+"lagoon-fail"
+    :: There is a vanishingly small chance of getting a NaN in a matrix which
+    :: collides with this, but it's like getting a hash table collision, very
+    :: very unlikely.  Also the Urbit NaN is hard-coded, `SINGNAN`.
+    =/  nan  0xffff.ffff
+    =/  p  (cantor m n)
+    `@lms`(not 5 +((mul m n)) (rap 5 ~[(fil 5 (mul m n) nan) (not 5 1 p)]))
+  ::
+  ::    Ones
+  ++  ones
+    ~/  %ones
+    |=  [m=@ud n=@ud]  ^-  @lms
+    ~_  leaf+"lagoon-fail"
+    =/  one  (bit:rs [%f s=%.y e=-0 a=1])
+    =/  not-one  (not 5 1 one)
+    =/  p  (cantor m n)
+    `@lms`(not 5 +((mul m n)) (rap 5 ~[(fil 5 (mul m n) not-one) (not 5 1 p)]))
+  ::
+  ::    Identity
+  ++  id
+    ~/  %id
+    |=  [m=@ud n=@ud]  ^-  @lms
+    ~_  leaf+"lagoon-fail"
+    =/  u  (zeros m n)
+    =/  i  1
+    |-
+      ?:  |(=(i m) =(i n))  u
+    $(i +(i), u (set u i i .1))  :: TODO test Hoon version (jet works)
```

Comment:
Changed to one arg; was two because of MATLAB syntax which allows asymmetric "identity" matrices.

`++repn` changed so this version doesn't work; I can munge the bits later but I'm marking this done for now.